library/Fission/Types.hs

```diff
@@ -413,18 +414,37 @@ instance User.Modifier Fission where
             zoneID     <- asks userZoneID
 
             let
-              subdomain = Just $ Subdomain rawUN
-              url       = URL {domainName, subdomain = Just (Subdomain "_did") <> subdomain}
-              did       = textDisplay (DID pk Key)
-              (_, didSegments) = Text.foldr splitter (0, ("" :| [])) did
+              subdomain   = Just $ Subdomain rawUN
+              url         = URL {domainName, subdomain = Just (Subdomain "_did") <> subdomain}
+              didSegments = Text.split256 $ DID.txtFromKey pk
 
             Route53.set Txt url zoneID didSegments <&> \case
               Left serverErr -> Error.openLeft serverErr
               Right _        -> Right pk
-    where
-      splitter :: Char -> (Natural, NonEmpty Text) -> (Natural, NonEmpty Text)
-      splitter chr (255, txtList)       = splitter chr (0, "" `cons` txtList)
-      splitter chr (len, (txt :| more)) = (len + 1, (Text.cons chr txt) :| more)
+
+  updateExchangeKeys uID keys now =
+    runDB (User.updateExchangeKeys uID keys now) >>= \case
```

Comment:
These 3 functions (`updatePublicKey`, `updateExchangeKeys` and `setData`) are all very similar, and I feel like we should be able to abstract out those similarities into a function. I've given it a go but I'm having issues with getting the types right on it :thinking:

Reply:
```haskell
runOperation ::
  ( MonadIO          m
  , MonadDB        t m
  , User.Retriever t
  )
  => m (Either User.Modifier.Errors Key.Public)
  -> UserId
  -> Text
  -> NonEmpty Text
  -> m (Either User.Modifier.Errors (NonEmpty Text))
runOperation dbFn uID subdomain txtSegments = 
  runDB dbFn >>= \case
    Left err ->
      return $ Left err
    Right _ -> do
      runDB (User.getById uID) >>= \case
        Nothing -> 
          return . Error.openLeft $ NotFound @User

        Just (Entity _ User { userUsername = Username rawUN }) -> do
          domainName <- asks userRootDomain
          zoneID     <- asks userZoneID

          let 
            url = URL {domainName, subdomain = Just (Subdomain subdomain) <> Just (Subdomain rawUN)}

          Route53.set Txt url zoneID txtSegments <&> \case
            Left serverErr -> Error.openLeft serverErr
            Right _        -> Right txtSegments
```
I think I'm missing something in the constraints

Reply:
Lemme take a look 👀

Reply:
Also, I'm a bit cautious in this specific case about making them all the same. You're right that the underlying implementation may be similar, but it feels brittle since they're _conceptually_ different effects.

The solution here may be to extract out a function that all 3 call. I'll play with the code you gave above in the meantime here, though.

Reply:
Working version:

```haskell

runOperation ::
     Fission (Either User.Modifier.Errors Key.Public)
  -> UserId
  -> Text
  -> NonEmpty Text
  -> Fission (Either User.Modifier.Errors (NonEmpty Text))
runOperation dbFn uID subdomain txtSegments =
  dbFn >>= \case
    Left err ->
      return $ Left err
  
    Right _ -> do
      runDB (User.getById uID) >>= \case
        Nothing ->
          return . Error.openLeft $ NotFound @User

        Just (Entity _ User { userUsername = Username rawUN }) -> do
          domainName <- asks userRootDomain
          zoneID     <- asks userZoneID

          let
            url = URL {domainName, subdomain = Just (Subdomain subdomain) <> Just (Subdomain rawUN)}

          Route53.set Txt url zoneID txtSegments <&> \case
            Left serverErr -> Error.openLeft serverErr
            Right _        -> Right txtSegments
```

Reply:
This probably seems odd. Where are the constraints? Well, in this scenario, we're making it concrete on `Fission`, _which we know_ is a `MonadIO`, `MonadDB`, and so on.

What happened to the `runDB`? Well, we probably want to just call the `User.getById` function or whatever. It calls `runDB` inside it, since we know the context it's going to get run in.

We also don't have to write an abstract version of this for the `Mock`. We just call this handy helper inside `Fission/Types.hs` and don't export it. We do this in a couple places (e.g. the big JWT module).

Reply:
In this implementation, I'd maybe even rename `dbFn` to `getPublicKey`, which looks like is the intent from the signature.

Also, we can probably tighten up which errors that part is allowed to return by changing to:

```haskell
Fission (Either User.Retriever.Errors Key.Public)
```

Which I believe is a subset of `User.Modifier.Errors`.

If we're always looking for the PK via the DB, we can even hardcode that DB call and drop the parameter.

Reply:
Nice this is perfect. exactly what I was shooting for!